pkg/ccl/changefeedccl/cdceval/functions.go

```diff
@@ -0,0 +1,183 @@
+			Fn:         prevRowAsJSON,
+			Info:       "Returns previous value of a row as JSONB",
+			Volatility: volatility.Stable,
+		}),
+}
+
+// TODO(yevgeniy): Implement additional functions (some ideas, not all should be implemented):
+//   * cdc_is_delete is easy; what about update? does update include new events?
+//   * cdc_is_new -- true if event is  a new row
+//   * tuple overload (or cdc_prev_tuple) to return previous value as a tuple
+//   * cdc_key -- effectively key_in_value where key columns returned as either a tuple or a json.
+//   * cdc_key_cols -- return key column names;
+//      * this can come in handy when working with jsonb; for example, emit previous JSONB excluding
+//        key columns can be done with `SELECT cdc_prev() - cdc_key_cols()
+//   * cdc_event_family_is(fam): return true if cdc event family is specified family; overload both for
+//     family ID and family name.
+//     function can be used to write complex conditionals when dealing with multi-family table(s)
+
+var cdcFnProps = &tree.FunctionProperties{
+	Category: "CDC builtin",
+}
+
+func makeCDCBuiltIn(fnName string, overloads ...tree.Overload) *tree.FunctionDefinition {
+	return tree.NewFunctionDefinition(fnName, cdcFnProps, overloads)
+}
+
+func cdcTimestampBuiltin(
+	fnName string, doc string, tsFn func(rowEvalCtx *rowEvalContext) hlc.Timestamp,
+) *tree.FunctionDefinition {
+	return tree.NewFunctionDefinition(
+		fnName,
+		cdcFnProps,
+		[]tree.Overload{
+			{
+				Types:      tree.ArgTypes{},
+				ReturnType: tree.FixedReturnType(types.Decimal),
+				Fn: func(evalCtx *eval.Context, datums tree.Datums) (tree.Datum, error) {
+					rowEvalCtx := rowEvalContextFromEvalContext(evalCtx)
+					return eval.TimestampToDecimalDatum(tsFn(rowEvalCtx)), nil
+				},
+				Info:       doc,
+				Volatility: volatility.Stable,
+			},
+		},
+	)
+}
+
+func prevRowAsJSON(evalCtx *eval.Context, _ tree.Datums) (tree.Datum, error) {
+	rec := rowEvalContextFromEvalContext(evalCtx)
+	if rec.memo.prevJSON != nil {
+		return rec.memo.prevJSON, nil
+	}
+
+	var prevJSON *tree.DJSON
+	if rec.prevRow.IsInitialized() {
+		b := jsonb.NewObjectBuilder(0)
+		if err := rec.prevRow.ForEachColumn().Datum(func(d tree.Datum, col cdcevent.ResultColumn) error {
+			j, err := tree.AsJSON(d, sessiondatapb.DataConversionConfig{}, time.UTC)
+			if err != nil {
+				return err
+			}
+			b.Add(col.Name, j)
+			return nil
+		}); err != nil {
+			return nil, err
+		}
+		prevJSON = tree.NewDJSON(b.Build())
+	} else {
+		prevJSON = tree.NewDJSON(jsonb.NullJSONValue)
+	}
+
+	rec.memo.prevJSON = prevJSON
+	evalCtx.Annotations.Set(cdcAnnotationAddr, rec)
+	return prevJSON, nil
+}
+
+type cdcCustomFunctionResolver struct {
+	sessiondata.SearchPath
+}
+
+// Resolve implements tree.CustomFunctionDefinitionResolver
+func (cdcCustomFunctionResolver) Resolve(name string) *tree.FunctionDefinition {
+	fn, found := cdcFunctions[name]
+	if found {
+		return fn
+	}
+	fn, found = cdcFunctions[strings.ToLower(name)]
```

Comment:
Would it make sense to do `cdcFunctions[name] = fn` if found so that we don't have to do this string call every time? Or is this resolution cached anyway?